cpp/src/fil/infer.cu

```diff
@@ -373,25 +382,30 @@ struct tree_aggregator_t<NITEMS, CATEGORICAL_LEAF> {
   }
 };
 
-template <int NITEMS, leaf_algo_t leaf_algo, class storage_type>
+template <int NITEMS, leaf_algo_t leaf_algo, bool cols_in_shmem,
+          class storage_type>
 __global__ void infer_k(storage_type forest, predict_params params) {
   extern __shared__ char smem[];
   float* sdata = (float*)smem;
+  int num_cols = params.ssp.num_cols;
   for (size_t block_row0 = blockIdx.x * NITEMS; block_row0 < params.num_rows;
        block_row0 += NITEMS * gridDim.x) {
-    // cache the row for all threads to reuse
-    for (size_t j = 0; j < NITEMS; ++j) {
-      size_t row = block_row0 + j;
+    size_t num_input_rows = min((size_t)NITEMS, params.num_rows - block_row0);
+    const float* block_input = params.data + block_row0 * num_cols;
+    if (cols_in_shmem) {
+      // cache the row for all threads to reuse
+      size_t feature;
```

Comment:
```suggestion
      size_t feature = 0;
```
To avoid uninitialized variables.

Reply:
Also `col` (as previously used) would be a better name for this variable, as it is more consistent with other names used in FIL (e.g. `num_cols`).